tun2socks/intra.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tun2socks
+
+import (
+	"errors"
+	"io"
+	"net"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-go-tun2socks/tun2socks/intra"
+	"github.com/eycorsican/go-tun2socks/core"
+)
+
+type intratunnel struct {
+	*tunnel
+	fakedns net.Addr
+	udpdns  net.Addr
+	tcpdns  net.Addr
+}
+
+// NewIntraTunnel creates a connected Intra session.
+//
+// `fakedns` is the DNS server (IP and port) that will be used by apps on the TUN device.
+//    This will normally be a reserved or remote IP address, port 53.
+// `udpdns` and `tcpdns` are the actual location of the DNS server in use.
+//    These will normally be localhost with a high-numbered port.
+func NewIntraTunnel(fakedns, udpdns, tcpdns string, tunWriter io.WriteCloser) (Tunnel, error) {
```

Comment:
There's a lot of logic here and in the tcp.go and udp.go files that are not Intra specific and are duplicated.

Instead, we could have a NewTunnel(dialer, tunWriter) method that can be reused, and all you need to do is implement the dialer interface.

The tunnel would take care of setting up the VPN, dialing and plugging everything together. This would be a nice change to upstream, as there is a lot of duplication in https://github.com/eycorsican/go-tun2socks/tree/master/proxy/socks too

Reply:
Discussed offline. The Dialer (or TCPDialer) approach can prevent the boilerplate and make the logic easier to test and reuse (in a commandline tool for example). But it's probably better to do it upstream later.